### ISSUE TITLE
How to synchronize after one epoch so that all ranks are having the same trained variables? 

### ISSUE THREAD

**Issue by wei-v-wang** [horovod/horovod#1006]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) TensorFlow
2. Framework version: TF v1.13.1
3. Horovod version: 0.13.1 & 0.16.1
4. MPI version: OpenMPI 3.0 & Intel MPI
5. CUDA version: NA
6. NCCL version: NA
7. Python version: 2.7.5
8. OS and version: CentOS

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
Similar issues to missing ranks (due to evaluation speed of the different ranks), may need to add barrier.

**Your question:**
Hi, I am using tf_cnn_benchmarks to train ResNet50 on 16 CPUs using horovod and the mode is train_and_eval. As tf_cnn_benchmarks is implemented currently, it seems after finishing the required steps (eval every N steps), each rank starts to evaluate on the validation dataset (50000 images). However they are output different evaluation results. I confirmed that all the ranks should be using the entire dataset. So it is likely that the ranks' trained variables are having different values (differ slightly) while doing evaluation. 
Accuracy @ 1 = 0.7455 Accuracy @ 5 = 0.9188 [50000 examples]
Accuracy @ 1 = 0.7459 Accuracy @ 5 = 0.9189 [50000 examples]
Accuracy @ 1 = 0.7453 Accuracy @ 5 = 0.9188 [50000 examples]
Accuracy @ 1 = 0.7459 Accuracy @ 5 = 0.9190 [50000 examples]
Accuracy @ 1 = 0.7468 Accuracy @ 5 = 0.9193 [50000 examples]
Accuracy @ 1 = 0.7458 Accuracy @ 5 = 0.9188 [50000 examples]
Accuracy @ 1 = 0.7456 Accuracy @ 5 = 0.9191 [50000 examples]
Accuracy @ 1 = 0.7450 Accuracy @ 5 = 0.9192 [50000 examples]
Accuracy @ 1 = 0.7456 Accuracy @ 5 = 0.9190 [50000 examples]
Accuracy @ 1 = 0.7459 Accuracy @ 5 = 0.9187 [50000 examples]
Accuracy @ 1 = 0.7458 Accuracy @ 5 = 0.9188 [50000 examples]
Accuracy @ 1 = 0.7458 Accuracy @ 5 = 0.9194 [50000 examples]
Accuracy @ 1 = 0.7460 Accuracy @ 5 = 0.9192 [50000 examples]
Accuracy @ 1 = 0.7460 Accuracy @ 5 = 0.9189 [50000 examples]
Accuracy @ 1 = 0.7422 Accuracy @ 5 = 0.9177 [50000 examples]
Accuracy @ 1 = 0.7458 Accuracy @ 5 = 0.9190 [50000 examples]
My question is how can I embed code to broadcast the trained variables and enforce a barrier as well? Thank you!

**Comment by alsrgv** [comment]
@wei-v-wang, sorry for the delayed response.  One simple way is to use `hvd.broadcast_global_variables(0)` operation at the end of the epoch, [like this](https://github.com/horovod/horovod/blob/master/examples/tensorflow_word2vec.py#L199).

**Comment by wei-v-wang** [comment]
Thanks @alsrgv! I will give it a try.  
The RN50 training algorithm seems to be only concerned with synchronizing (allreduce) gradients, so having different trained weights for each worker might be the expected behaviour during a multi-node training. It's just that without a explicit barrier, horovod produced warnings like "missing ranks.....", but RN50 still converged in the end. I read from another discussion thread that it is expected, especially when a new epoch starts, and when different ranks have different speed in inference and finish in different amount of time.  I realize there is a flag to turn off the warning message. 
My question is should I still be confident with the training results if the logs containing "missing ranks" but the evaluation results show good accuracy? Thank you!

**Comment by alsrgv** [comment]
@wei-v-wang, yes - missing ranks just indicates inefficient usage of resources, but does not indicate any numerical issues.

**Comment by wei-v-wang** [comment]
Ok, thanks Alex @alsrgv for the assurance. I think this issue can be closed now. I created this issue mainly for getting an answer for the above missing rank question, which just got answered. Thank you!
